File: bnf/src/parse.ts
import antlr4 from 'antlr4';
import DspLexer from './generated/DspLexer.js';
import DspParser from './generated/DspParser.js';
import DspVisitor from './generated/DspVisitor.js';

/**
 * Visitor to convert from traditional Dsp syntax to lisp-like Ds syntax
 */
class ParseVisitor extends DspVisitor {
    visitRule_pool(ctx) {
        const rules = ctx.rule_();
        if (!rules || rules.length === 0) {
            return '';
        }
        return rules.map(r => this.visit(r)).join('\n');
    }

    visitRule(ctx) {
        const terms = ctx.term();
        if (!terms || terms.length === 0) {
            return '';
        }
        
        const result = terms.map(t => this.visit(t));
        
        // Check if this is a rule with arrow (->)
        const text = ctx.getText();
        if (text.includes('->')) {
            // Multiple premises with conclusion
            const conclusion = result.pop();
            return result.join('\n') + '\n----------\n' + conclusion;
        } else {
            // Just a fact (single term)
            return result[0];
        }
    }

    visitSymbol(ctx) {
        return ctx.SYMBOL().getText();
    }

    visitParentheses(ctx) {
        return this.visit(ctx.term(0));
    }

    visitSubscript(ctx) {
        const terms = ctx.term();
        const base = this.visit(terms[0]);
        const indices = terms.slice(1).map(t => this.visit(t));
        return `(subscript ${base} ${indices.join(' ')})`;
    }

    visitFunction(ctx) {
        const terms = ctx.term();
        const func = this.visit(terms[0]);
        const args = terms.slice(1).map(t => this.visit(t));
        
        if (args.length === 0) {
            return `(function ${func})`;
        }
        return `(function ${func} ${args.join(' ')})`;
    }

    visitUnary(ctx) {
        const op = ctx.getChild(0).getText();
        const operand = this.visit(ctx.term(0));
        return `(unary ${op} ${operand})`;
    }

    visitBinary(ctx) {
        const terms = ctx.term();
        const left = this.visit(terms[0]);
        const right = this.visit(terms[1]);
        
        // Find the operator - it's the token between the two terms
        // Iterate through children to find terminal nodes (operators)
        let op = '';
        for (let i = 0; i < ctx.getChildCount(); i++) {
            const child = ctx.getChild(i);
            // Check if this is a terminal node (not a term context)
            if (!child.term && child.symbol) {
                op = child.getText();
                break;
            } else if (typeof child.getText === 'function') {
                const text = child.getText();
                // Skip if it matches term outputs
                if (text !== left && text !== right && !text.includes('(') && text.length > 0) {
                    op = text;
                    break;
                }
            }
        }
        
        return `(binary ${op} ${left} ${right})`;
    }
}

/**
 * Convert from traditional Dsp syntax to lisp-like Ds syntax
 * @param {string} input - Input text in Dsp syntax
 * @returns {string} Output text in Ds syntax
 */
export function parse(input) {
    const chars = new antlr4.CharStream(input);
    const lexer = new DspLexer(chars);
    const tokens = new antlr4.CommonTokenStream(lexer);
    const parser = new DspParser(tokens);
    const tree = parser.rule_pool();
    const visitor = new ParseVisitor();
    return visitor.visit(tree);
}

File: bnf/apyds_bnf/parse.py
"""
Parse: Convert from traditional Dsp syntax to lisp-like Ds syntax
"""

from antlr4 import InputStream, CommonTokenStream
from .generated.DspLexer import DspLexer
from .generated.DspParser import DspParser
from .generated.DspVisitor import DspVisitor


class ParseVisitor(DspVisitor):
    """Visitor to convert from traditional Dsp syntax to lisp-like Ds syntax"""

    def visitRule_pool(self, ctx):
        rules = ctx.rule_()
        if not rules:
            return ""
        return "\n".join(self.visit(r) for r in rules)

    def visitRule(self, ctx):
        terms = ctx.term()
        if not terms:
            return ""

        result = [self.visit(t) for t in terms]

        # Check if this is a rule with arrow (->)
        text = ctx.getText()
        if "->" in text:
            # Multiple premises with conclusion
            conclusion = result.pop()
            return "\n".join(result) + "\n----------\n" + conclusion
        else:
            # Just a fact (single term)
            return result[0]

    def visitSymbol(self, ctx):
        return ctx.SYMBOL().getText()

    def visitParentheses(self, ctx):
        return self.visit(ctx.term(0))

    def visitSubscript(self, ctx):
        terms = ctx.term()
        base = self.visit(terms[0])
        indices = " ".join(self.visit(t) for t in terms[1:])
        return f"(subscript {base} {indices})"

    def visitFunction(self, ctx):
        terms = ctx.term()
        func = self.visit(terms[0])
        args = " ".join(self.visit(t) for t in terms[1:])

        if not args:
            return f"(function {func})"
        return f"(function {func} {args})"

    def visitUnary(self, ctx):
        op = ctx.getChild(0).getText()
        operand = self.visit(ctx.term(0))
        return f"(unary {op} {operand})"

    def visitBinary(self, ctx):
        terms = ctx.term()
        left = self.visit(terms[0])
        right = self.visit(terms[1])

        # Find the operator - it's the token between the two terms
        # Iterate through children to find terminal nodes (operators)
        op = ""
        for i in range(ctx.getChildCount()):
            child = ctx.getChild(i)
            # Check if this is a terminal node (has a symbol attribute)
            if hasattr(child, 'symbol'):
                op = child.getText()
                break
            else:
                text = child.getText()
                # Skip if it matches term outputs
                if text != left and text != right and "(" not in text and len(text) > 0:
                    op = text
                    break

        return f"(binary {op} {left} {right})"


def parse(input_text: str) -> str:
    """
    Convert from traditional Dsp syntax to lisp-like Ds syntax

    Args:
        input_text: Input text in Dsp syntax

    Returns:
        Output text in Ds syntax
    """
    input_stream = InputStream(input_text)
    lexer = DspLexer(input_stream)
    token_stream = CommonTokenStream(lexer)
    parser = DspParser(token_stream)
    tree = parser.rule_pool()
    visitor = ParseVisitor()
    return visitor.visit(tree)
